### PR TITLE
Replace setup.py with pyproject.toml

### PR DESCRIPTION
**Pull request by skirpichev**
FYI: I see 'DEPRECATION: flake8-sfs is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change.', hence this pr.

### CODE REVIEW

**Comment by peterjc** [comment]
It looks like we had no v0.0.3 wheel on PyPI, worth rectifying: https://github.com/peterjc/flake8-sfs/commit/3f824f9e585d808f7c08b001ea445f3531811123

Anyway, the pip team are very clear about the way things are going, e.g. https://pip.pypa.io/en/stable/reference/build-system/ so the change from setup.py with pyproject.toml will need to happen.

File: setup.py [hunk 1]
```diff
@@ -1,7 +1,5 @@
 """Setup file for building/installing flake8-sfs."""
 
-from __future__ import with_statement
```
Comment (peterjc) [inline setup.py:3]: Not useless, it was important to ensure ``python2 setup.py ...`` works as well as possible. i.e. Clear error message, not an ugly exception.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+classifiers = ['Development Status :: 4 - Beta',
+               'Intended Audience :: Developers',
+               'Framework :: Flake8',
+               'License :: OSI Approved :: MIT License',
+               'Operating System :: OS Independent',
+               'Programming Language :: Python',
+               'Topic :: Software Development :: Libraries :: Python Modules',
+               'Topic :: Software Development :: Libraries :: Quality Assurance',
+               'Programming Language :: Python :: 3',
+               'Programming Language :: Python :: 3 :: Only']
```
Comment (peterjc) [inline pyproject.toml]: I would use black-style indentation here with each entry on its own line like https://peps.python.org/pep-0621/#example

**Comment by peterjc** [comment]
Did you do the conversion to ``pyproject.toml`` by hand, drawing on your existing knowledge? Or is there an off-the-shelf tool to an automated starting point?

e.g. The dynamic version stuff looked tricky, explained here: https://github.com/pypa/setuptools/discussions/3190 and probably in the documentation somewhere as well.

**Comment by peterjc** [comment]
The build system is failing as it still tried to run ``setup.py``. I have seen some people leave a stub ``setup.py`` in place as follows:

```python
#!/usr/bin/env python
import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

Maybe we should too, and/or update the GitHub Actions install command - probably to just ``pip install .`` instead?

File: setup.py (hunk 1 above)
Comment (skirpichev) [reply setup.py:3]: Well, if you require python >= 3.7 and think that the exception is ugly - there should be a bug somewhere else.

Comment (peterjc) [reply setup.py:3]: Less and less important, I grant you - but that's why the future-import was there ;)

**Comment by skirpichev** [comment]
> Did you do the conversion to `pyproject.toml` by hand, drawing on your existing knowledge?

Yes.

> The build system is failing as it still tried to run setup.py.

May fault.  I will try to switch things to be like `pip install .`.

> I have seen some people leave a stub setup.py in place as follows

That depends on your wishes to support old versions of setuptools.

**Comment by peterjc** [comment]
I still have "muscle memory" for ``python setup.py install``, still retraining myself to use ``pip install .`` instead ;)

I was hoping automatic conversion had matured somewhat, as there are half a dozen projects of varying size I anticipate having to migrate to ``pyproject.toml`` eventually. Happily two are also pure python flake8 plugins so this would be a very good reference point.

This change seems enough to justify rolling out v0.0.4, would you like to be acknowledged by name in the change log? See e.g. entries in https://github.com/peterjc/flake8-black/blob/master/README.rst or https://github.com/peterjc/flake8-rst-docstrings/blob/master/README.rst

**Comment by skirpichev** [comment]
> I was hoping automatic conversion had matured somewhat

I don't know a mature tool, unfortunately.

> there are half a dozen projects of varying size I anticipate having to migrate to pyproject.toml eventually

flake8 itself is a most famous example:)  The point is you can't use pip after 23.1...

> This change seems enough to justify rolling out v0.0.4

Maybe.

> would you like to be acknowledged by name in the change log?

It's up to you how you acknowledge contributors.

**Comment by peterjc** [comment]
I've found on larger projects that not everyone wants their name listed as a contributor. I've not set out any expectations in a CONTRIBUTING file or pull request template here. I take it you'd be happy to be named though.

**Comment by peterjc** [comment]
It took a while to find out what to replace my old commands to make the source tar-ball and wheel with (quoting my README file):

```
python setup.py sdist --formats=gztar && python setup.py bdist_wheel
```

There isn't currently an sdist command replacement in pip. Skimming various discussions and issues, it looks like for now the best choice is https://github.com/pypa/build which is just one command which defaults to using the current directory as input, and builds both the tar-ball and the wheel from it:

```
python -m build
```

(I have not tried this on a more complex example, e.g. with C code or complicated dependencies)

My notes in the README already assumed ``twine`` was installed, but would need ``pip install build twine`` now.

Would you use that (pypa/build), or something else?

**Comment by skirpichev** [comment]
> I've not set out any expectations in a CONTRIBUTING file or pull request template here. I take it you'd be happy to be named though.

Feel free to do as you like.

> My notes in the README already assumed twine was installed, but would need pip install build twine now.

I would suggest to add optional dependencies in the pyproject.toml (i.e. devel, tests and so on).

> Would you use that (pypa/build), or something else?

Yes, I use pypa/build for my projects, e.g. you could take look on the [diofant](https://github.com/diofant/diofant/) and it's GH workflow.  At least for pure-Python projects this seems to be a good option.

**Comment by skirpichev** [comment]
BTW, you could use [gh-action-pypi-publish](https://github.com/pypa/gh-action-pypi-publish) to automate PyPI uploads from your workflow.

**Comment by peterjc** [comment]
Applied and released as v0.0.4, almost painless - a classifier had changed: https://github.com/peterjc/flake8-sfs/commit/6ab5d074a04a82948f2d942cc7159482d17c0ec8

Thank you this, I'll apply some of the lessons here to other projects - although those with C code will need more research.

**Pull request merged**